### PR TITLE
Audit log blog post 

### PR DESCRIPTION
**Pull request by gunnarmorling**
Blog post accompanying the [audit log demo](https://github.com/debezium/debezium-examples/tree/master/auditlog). Feedback welcome!

### CODE REVIEW

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+Traditionally, application-side libraries and database triggers have been common means for creating audit trails:
+when a data record is inserted, updated or deleted, they'd maintain specific column(s) in the data tables ("createdBy", "lastUpdated" etc.) and/or copy earlier record versions into some form of history tables.
+This comes with some disadvantages, though:
+
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction and thus may cause longer response times of the application
+
+Library-based solutions also oftentimes cannot provide metadata in case of bulk updates and deletes (`UPDATE mytable WHERE ...`),
+as the listeners used to hook them into the persistence framework are not aware of all the affected records.
+Also, by definition, any approach based on an application-side library is prone to miss any data changes done directly in the database, e.g. when running a data load or bypassing the application during an emergency data patch.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Also here the same holds true for the proposed solution. Since direct changes won't trigger application level meta data insertion which is done in the example code. The only difference is that you can e.g. "block the enrichments" in the stream processor but still there is no such data available under such circumstances.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+
+These problems don't exist when leveraging the transaction log as the source for an audit trail and using change data capture for retrieving the change information and sending it into a message broker or log such as Apache Kafka.
+Running asynchronously, the CDC process can extract the change data without impacting OLTP transactions, not missing any event, no matter whether what its origin is and whether it was caused by a bulk operation.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: ...,no matter ~whether~ what its origin and whether ...

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+        vegetable = vegetableService.createVegetable(vegetable);
+
+        return Response.ok(vegetable).status(Status.CREATED).build();
+    }
+
+    // update, delete ...
+}
+----
+
+Not too exciting, if you've ever built REST services with JAX-RS before:
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: I'd rephrase that somehow even though you are right about the "excitment" this might give readers :-)

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+----
+vegetablesdb> select * from inventory.vegetable;
++------+---------------+---------+
+| id   | description   | name    |
+|------+---------------+---------|
+| 1    | Spicy         | Potatoe |
+| 11   | Delicious!    | Banana  |
+| 10   | tasty         | Pear    |
++------+---------------+---------+
+SELECT 3
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: I would suggest to remove this info about rows + timing. WDYT?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+Time: 0.026s
+----
+
+== Enriching Change Events with Auditing Metadata
+
+With the business data (vegetables) and the transaction-scoped metadata being stored in the database,
+it's time to set up the link:/documentation/reference/0.10/connectors/postgresql.html[Debezium Postgres connector] and stream the data changes from the `vegetable` and `transaction_context_data` tables into corresponding Kafka topics.
+Again refer to the example README file for the details of https://github.com/debezium/debezium-examples/tree/master/auditlog#deploy-the-debezium-postgres-connector[deploying the connector].
+
+The `dbserver1.inventory.vegetable` topic should contain contain change events for created, updated and deleted vegetable records, whereas the `dbserver1.inventory.transaction_context_data` topic should only contain create messages for each inserted metadata record.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: "contain" 2x

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 6]
```diff
@@ -0,0 +1,647 @@
...
+the `client_date`, `usecase` and `user_name` attributes from the former can be added to the latter:
+
+++++
+<div class="imageblock centered-image">
+    <img src="/images/auditing_output_message.png" style="max-width:100%; margin-bottom:10px; margin-top:10px;" class="responsive-image" alt="Enriched Vegetable Message">
+</div>
+++++
+
+This kind of message transformation is a perfect use case for https://kafka.apache.org/documentation/streams/[Kafka Streams],
+a Java API for implementing stream processing applications on top of on Kafka topics,
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: ...on top of ~on~ Kafka topics,

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+
+When thinking about the actual implementation of the enrichment logic,
+a https://kafka.apache.org/23/documentation/streams/developer-guide/dsl-api.html#kstream-kstream-join[stream-to-stream] join might appear as a suitable solution.
+By creating ``KStream``s for the two topics, we may try and implement the joining functionality.
+This creates a problem though in regards to ordering guarantees of the change events.
+By default, Debezium will use a table's primary key as the message key for the corresponding Kafka messages.
+This means that all messages for the same vegetable record will have the same key and thus will go into the same partition of the vegetables Kafka topic.
+This in turn guarantees that a consumer of these events sees all the messages pertaining to the same vegetable record in the exact same order as they were created.
+
+Now, in order to join the two streams, the message key must be the same on both sides.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: You might add just 1 sentence that a stream-stream join would also need a window for a length which is not easy to define in general.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 8]
```diff
@@ -0,0 +1,647 @@
...
+}
+----
+
+The `StreamsBuilder` from Kafka Streams is used here to
+
+* set up a state store which will serve as the buffer for change events that cannot be processed yet
+* define a `GlobalKTable` based on the transaction metadata topic and
+* define a `KStream` based on the vegetables topic; on this stream, any incoming tombstone markers are filtered, the reasoning being that the retention policy for an audit trail topic typically should be time-based than based on log compaction; similarly, snapshot events are filtered, assuming they are not relevant for an audit trail and there wouldn't be any corresponding metadata provided by the application for the snapshot transaction initiated by the Debezium connector
++
+Any other messages are enriched with the corresponding transaction metadata via a custom `Transformer` (see below) and finally are written to an output topic
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: and finally ~are~ written

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 9]
```diff
@@ -0,0 +1,647 @@
...
+
+    @Override
+    public void close() {
+    }
+}
+----
+
+The basic logic of the transformer is this:
+
+* When a vegetables change event arrives, look up the corresponding metadata in the statestore of the
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: statestore -> state store

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 10]
```diff
@@ -0,0 +1,647 @@
...
+The basic logic of the transformer is this:
+
+* When a vegetables change event arrives, look up the corresponding metadata in the statestore of the 
+transaction topic's `GlobalKTable`, using the transaction id from the `source` block of the change event as the key
+* If the metadata could be found, add the metadata to change event (under the `audit` field) and return that enriched event
+* If the metadata could not be found, add the incoming event into the buffer of change events and return
+* Before actually getting to the incoming event, all buffered events are processed; only if all could be enriched, the incoming event will be processed, too
+* In order to emit buffered events also if no new change event is coming in, the `punctuate()` method is used to periodically process the buffer
+
+The key piece is the buffer for unprocesseable change events.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: unprocess-e-able

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 11]
```diff
@@ -0,0 +1,647 @@
...
+To maintain the order of events, the buffer must be processed in order of insertion,
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: FIFO might help to add to the sentence somewhere?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 12]
```diff
@@ -0,0 +1,647 @@
...
+kafka-streams.default.key.serde=io.debezium.demos.auditing.enricher.JsonObjectSerde
+kafka-streams.default.value.serde=io.debezium.demos.auditing.enricher.JsonObjectSerde
+kafka-streams.processing.guarantee=exactly_once
+----
+
+In the next step let's take a look at the `ChangeEventEnricher` class, our custom transformer.
+The implemention is based on the assumption that change events are serialized as JSON,
+but of course it could be done equally well using other formats such as Avro or Protocol Buffers.
+
+This is a bit of code, but hopefully its decomposition into multiple smaller methods makes it comprehensible:
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Since you describe only the most relevant parts below the code block maybe there is not need to provide all code lines here but instead reduce it to the code blocks that are also discussed in the text and link to the source file. WDYT?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 13]
```diff
@@ -0,0 +1,647 @@
...
+Additional metadata that typically should be part of an audit log,
+can be provided by the application via a separate table,
+which also is captured via Debezium.
+With the help of Kafka Streams the actual data change events can be enriched with the data from that metadata table.
+
+One aspect we haven't discussed yet is querying the audit trail entries,
+e.g. to examine specific earlier versions of the data.
+To do so, the enriched change data events typically would be stored in a queryable database.
+Unlike a basic data replication pipeline, not only the latest version of each record would be stored in the database in that case, but all the versions, i.e. the primary keys typically would be amended with the transaction id of each change.
+This would all to select single data records or even joins of multiple tables to get the data valid as per a given transaction id.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: all -> allow

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 14]
```diff
@@ -0,0 +1,647 @@
...
+How this could be implemented in detail, may be discussed in a future post.
```
Comment (hpgrahsl) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: no comma needed I guess?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 15]
```diff
@@ -0,0 +1,647 @@
...
+Reasons for maintaining data audit trails are manifold:
+e.g. regulatory requirements may mandate businesses to keep complete historic information of their customer, purchase order, invoice or other data.
+Also for an enterprise's own purposes it can be very useful to be able to analyze why and how certain data has changed, e.g. allowing to improve business processes and analyze errors.
+
+Traditionally, application-side libraries and database triggers have been common means for creating audit trails:
+when a data record is inserted, updated or deleted, they'd maintain specific column(s) in the data tables ("createdBy", "lastUpdated" etc.) and/or copy earlier record versions into some form of history tables.
+This comes with some disadvantages, though:
+
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction and thus may cause longer response times of the application
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: The difference is in the number of records; indeed some kind of metadata record will be written in both cases. But a library like Envers or a triggers based approach will *also* write the entries in the history table: for each update and delete there'd be another insert in the history table. Whereas this is not the case for the proposed CDC-based solution, where that information is obtained from the log.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 1 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Yes, that's true as far as the metadata information is concerned. The actual change events of course would be captured from the log also for a bulk update/delete. I'll see how I can rephrase that; I was unsure about that myself, but thought I'd get away with it. Seems that not :)

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 12 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Yes, I was wondering that, too. But I've added AsciiDoc callouts now, which I think help a lot with keeping this comprehensible (hard to see here on GitHub, but they'll show up in the rendered output on the website). So I think it's ok as is. I've already removed the `BufferOffsets` class and just point to it on GitHub.

Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Here's an image of the rendered callouts; they're referenced in text below the listing. Screen Shot 2019-09-29 at 22 14 43

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 15 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: I've added a sentence to make it clearer:

 >writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction (for each update or delete, also an insert must be written into the corresponding history table) and thus may cause longer response times of the application

Does that work better?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 3 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Yes, good point. It didn't feel quite right to me, too, but I couldn't quite say what it is. Will reword.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 8 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Hum, I think this was correct?

**Comment by gunnarmorling** [comment]
Thanks a lot for the thorough proof-reading, @hpgrahsl! I've pushed one more commit addressing most your remarks and resolved the corresponding discussions.

**Comment by jbfletch** [comment]
Still reading but one comment is that all your Farmers in the transactions table are men as a lady farmer it would be nice to see a Farmer Margaret 🙂

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+| id   | description   | name    |
+|------+---------------+---------|
+| 1    | Spicy         | Potatoe |
+| 11   | Delicious!    | Banana  |
+| 10   | tasty         | Pear    |
++------+---------------+---------+
+----
+
+[source,sql]
+----
```
Comment (jpechane) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Minor nit pick, why in the example above in the trasactions table the usernames are with space and below without?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 17]
```diff
@@ -0,0 +1,656 @@
...
+<div class="imageblock centered-image">
+    <img src="/images/auditing_output_message.png" style="max-width:100%; margin-bottom:10px; margin-top:10px;" class="responsive-image" alt="Enriched Vegetable Message">
+</div>
+++++
+
+This kind of message transformation is a perfect use case for https://kafka.apache.org/documentation/streams/[Kafka Streams],
+a Java API for implementing stream processing applications on top of Kafka topics,
+providing operators that let you filter, transform, aggregate and join Kafka messages.
+
+As foundation for our stream processing application we're going to use https://quarkus.io/[Quarkus],
```
Comment (jpechane) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Is not runtime environment better word than foundaiton in this context?

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 18]
```diff
@@ -0,0 +1,656 @@
...
+    @Override
+    public void close() {
+    }
+}
+----
+<1> When a vegetables change event arrives, look up the corresponding metadata in the state store of the
+transaction topic's `GlobalKTable`, using the transaction id from the `source` block of the change event as the key;
+if the metadata could be found, add the metadata to change event (under the `audit` field) and return that enriched event
+<2> If the metadata could not be found, add the incoming event into the buffer of change events and return
+<3> Before actually getting to the incoming event, all buffered events are processed; only if all could be enriched, the incoming event will be processed, too
```
Comment (jpechane) [inline blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: You should explain why

**Comment by gunnarmorling** [comment]
> Still reading but one comment is that all your Farmers in the transactions table are men as a lady farmer it would be nice to see a Farmer Margaret 

@jbfletch, uff, absolutely, will update. Sorry Barry, Margaret it is :)

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 16 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Good catch; Updating the image.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 19]
```diff
@@ -0,0 +1,656 @@
+= Building Audit Logs with Change Data Capture and Stream Processing
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Yes, will do. Was only waiting with it until the end, as I didn't want to mess up with the review comments.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 7 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Done.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc [hunk 20]
```diff
@@ -0,0 +1,656 @@
...
+Also for an enterprise's own purposes it can be very useful to be able to analyze why and how certain data has changed, e.g. allowing to improve business processes and analyze errors.
+
+Traditionally, application-side libraries and database triggers have been common means for creating audit trails:
+when a data record is inserted, updated or deleted, they'd maintain specific column(s) in the data tables ("createdBy", "lastUpdated" etc.) and/or copy earlier record versions into some form of history tables.
+This comes with some disadvantages, though:
+
+* metadata columns and triggers add bloat to the model in the database and the corresponding application domain model
+* writing records in history tables as part of OLTP transactions increases the number of executed statements within the transaction (for each update or delete, also an insert must be written into the corresponding history table) and thus may cause longer response times of the application
+
+Library-based solutions also oftentimes cannot provide metadata in case of bulk updates and deletes (`UPDATE mytable WHERE ...`),
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: Clarified that the TX log will contain one entry for each record updated/deleted in a bulk operation.

**Comment by gunnarmorling** [comment]
Pushed one more commit to address Jiri's remarks and add Margaret :)

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 15 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: I've largely restructured this section. It should be much clearer now.

File: blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc (hunk 1 above)
Comment (gunnarmorling) [reply blog/2019-09-12-audit-logs-with-change-data-capture-and-kafka-streams.adoc]: I hope it's clearer now with the rewording of this section.

**Comment by gunnarmorling** [comment]
Thanks a lot for all the feedback, everyone! I think I've incorporated all the suggestion and am quite happy about this piece. I'll push it to the website later today. Thanks again!

**Comment by gunnarmorling** [comment]
And it's live: https://twitter.com/gunnarmorling/status/1179062410134638594 :)
